operator/bundle_helpers/patch-csv.py

```diff
@@ -69,33 +79,64 @@ def patch_csv(csv_doc, version, operator_image, first_version, no_related_images
     if rbac_proxy_replacement:
         rewrite(csv_doc, rbac_proxy_replace(rbac_proxy_replacement))
 
-    x, y, z = (int(c) for c in version.split('-', maxsplit=1)[0].split('.'))
-    first_x, first_y, first_z = (int(c) for c in first_version.split('-', maxsplit=1)[0].split('.'))
     previous_y_stream = get_previous_y_stream(version)
 
     # An olm.skipRange doesn't hurt if it references non-existing versions.
     csv_doc["metadata"]["annotations"]["olm.skipRange"] = f'>= {previous_y_stream} < {version}'
 
-    if version_skips:
-        csv_doc["spec"]["skips"] = version_skips
-
     # multi-arch
     if "labels" not in csv_doc["metadata"]:
         csv_doc["metadata"]["labels"] = {}
     for arch in extra_supported_arches:
         csv_doc["metadata"]["labels"][f"operatorframework.io/arch.{arch}"] = "supported"
 
-    if (x, y, z) > (first_x, first_y, first_z):
-        if z == 0:
-            csv_doc["spec"]["replaces"] = f'{raw_name}.v{previous_y_stream}'
-        else:
-            csv_doc["spec"]["replaces"] = f'{raw_name}.v{x}.{y}.{z - 1}'
+    skips = parse_skips(csv_doc["spec"], raw_name)
+    replaced_xyz = calculate_replaced_version(
+        version=version, first_version=first_version, previous_y_stream=previous_y_stream, skips=skips)
+    if replaced_xyz is not None:
+        csv_doc["spec"]["replaces"] = f"{raw_name}.v{replaced_xyz}"
 
     # OSBS fills relatedImages therefore we must not provide that ourselves.
     # Ref https://osbs.readthedocs.io/en/latest/users.html?highlight=relatedImages#creating-the-relatedimages-section
     del csv_doc['spec']['relatedImages']
 
 
+def parse_skips(spec, raw_name):
+    raw_skips = spec.get("skips", [])
+    return set([XyzVersion.parse_from(must_strip_prefix(item, f"{raw_name}.v")) for item in raw_skips])
+
+
+def must_strip_prefix(str, prefix):
+    if not str.startswith(prefix):
+        raise RuntimeError(f"{str} does not begin with {prefix}")
+    return str[len(prefix):]
+
+
+def calculate_replaced_version(version, first_version, previous_y_stream, skips):
+    current_xyz = XyzVersion.parse_from(version)
+    first_xyz = XyzVersion.parse_from(first_version)
+    previous_xyz = XyzVersion.parse_from(previous_y_stream)
+
+    if current_xyz <= first_xyz:
+        return None
+
+    initial_replace = previous_xyz if current_xyz.z == 0 else \
+        XyzVersion(current_xyz.x, current_xyz.y, current_xyz.z - 1)
+
+    current_replace = initial_replace
+
+    while current_replace in skips:
+        logging.info(f"Looks like {current_replace} replace version is in skips list, trying next patch.")
+        current_replace = XyzVersion(current_replace.x, current_replace.y, current_replace.z + 1)
```

Comment:
There's a corner case here that will produce a wrong output, however it's quite unlikely that we'd ever have such a case.

Let's say that we're releasing 4.2.0 and want to skip all all 4.1.x versions released so far (latest is 4.1.3). The algorithm will then output 4.1.4, which does not exist (and we cannot know that, since this script doesn't know which versions actually exist).

Reply:
I agree. It's both possible and the likelihood is low. If Y-Stream that's getting `skipped` is still supported, we should make a patch with fixes because it's not expected to leave supported Y-Stream broken.

Reply:
Since GitHub forces to resolve conversations before merging, I added some code comments about this to be able to feel happy when clicking `Resolve conversation` button.